compiler/rustc_span/src/span_encoding.rs

```diff
@@ -87,45 +87,130 @@ pub struct Span {
     ctxt_or_parent_or_marker: u16,
 }
 
-impl Span {
+// Convenience structures for all span formats.
+#[derive(Clone, Copy)]
+struct InlineCtxt {
```

Comment:
Since these are being transmuted, can you make these structs `repr(C)` so that the compiler can't start reordering fields in memory?

Reply:
Adding `repr(C)` here would seem like an excessive pedantry to me.
Compiler shouldn't perform reordering if it's not beneficial (and it's not beneficial here), and all structures have the same layout so even a relatively unreasonable compiler would reorder them identically, and I don't want to add repr because it may affect (?) how `Span` is passed between functions.
Also we *are* the compiler, so we know what the compiler will do.

Reply:
AFAIK, some safe transmute features are in progress, ideally safe use cases like this (~name aliases for fields) would use something like that.

Reply:
> Adding repr(C) here would seem like an excessive pedantry to me.

It does? It seems like standard procedure to me. I agree that it's unlikely that the compiler will choose a different field order for some of these structs, but technically it is free to do so and the reference [explicitly suggests](https://doc.rust-lang.org/reference/type-layout.html#the-c-representation) using `repr(C)` when reinterpreting values as different types. I think it's fine to be pedantic when transmute is involved.

Reply:
I've just removed the transmutes, they are not strictly necessary anymore because `match_span_kind` works by value.
It's more code, but at least there's no unsafe and no opportunity repr arguments.

Reply:
Good to hear the transmutes are gone. The argument that the compiler should or would do particular things on non-`repr(C)` types was making me nervous!

Reply:
> I've just removed the transmutes

Thanks! That's even better.